lib/src/pubspec.dart

```diff
@@ -29,13 +30,53 @@ final _packageName = new RegExp(
 /// The default SDK upper bound constraint for packages that don't declare one.
 ///
 /// This provides a sane default for packages that don't have an upper bound.
-final _defaultUpperBoundSdkConstraint = new VersionConstraint.parse("<2.0.0");
+final VersionRange _defaultUpperBoundSdkConstraint =
+    new VersionConstraint.parse("<2.0.0");
+
+/// The upper bound contraint that matches the dev sdk.
+final _preReleaseTwoDotZeroSdkVersion = new Version.parse("2.0.0-dev.infinity");
+
+/// Whether or not to allow the pre-release sdk for packages that have an
+/// upper bound Dart SDK constraint of <2.0.0.
+///
+/// If enabled then a Dart SDK upper bound of <2.0.0 is always converted to
+/// <2.0.0-dev.infinity.
+///
+/// This has a default value of `true` but can be overridden with the
+/// PUB_ALLOW_PRERELEASE_SDK system environment variable.
+final bool allowPreReleaseTwoDotZeroSdk = () {
+  var userSetting =
+      Platform.environment["PUB_ALLOW_PRERELEASE_SDK"]?.toLowerCase() ?? 'true';
+  switch (userSetting) {
+    case "true":
+      return true;
+      break;
+    case "quiet":
+      warnAboutPreReleaseTwoDotZeroSdkOverrides = false;
```

Comment:
PM nit...

This makes me sad. I hate mutable top-level things.

Could their be a private field w/ an enum type and then two public bool properties that just key off it?

Reply:
done (didn't use an enum though)